### ISSUE TITLE
Problem with integer normalization?

### ISSUE THREAD

**Issue by stefanocereda** [scikit-optimize/scikit-optimize#503]
Hi all, I'm doing optimization on a problem with a lot of integer variables and I've observed much better performance when using spearmint instead of skopt. Since I really prefer skopt interface I would like to replicate spearmint's performance. After some testing I think I may have some problems with the normalization of integer values.

I've run some tests on the branin function varying the domains:
1) original branin function with domain [(-5.0, 10.0), (0.0, 15.0)]
2) branin scaled with domain [(-5.0, 10.0), (0.0, 15000.0)]
where by branin scaled I mean:
```
def branin_scaled(xs):
    xs = [xs[0], xs[1]/1000]
    return branin(xs)
```
3) branin scaled with domain [(-5.0, 10.0), (0, 15000)]

What I would expect is:
1-2 should be pretty similar: the actual domain is the same but skopt should see different scales, leading to different kernel behavior.
2-3 should also be pretty similar: it's true that skopt only sees integer numbers, but by dividing it I go back to a number with 3 decimal digits, which I think should be enough to optimize the function.

Instead, I've observed a much worse convergence on 3.
Specifically, I've run the tests for 100 epochs and repeated them for 50 runs, here's the graph of the average value obtained at each epoch:

![image](https://user-images.githubusercontent.com/11663239/30016541-de119c30-9155-11e7-914a-366369df08d5.png)

(I've also included a test with a much bigger domain and the original branin function, obtaining a reasonably bad result. However, this is still better than scenario 3)

Is this behavior expected or is there some problem with integer normalization?

**Comment by stefanocereda** [comment]
I've made another test where the variable is divided in a wrong way. I use as domain [(-5.0, 10.0), (0.0, 15000.0)] and the second variable is divided by 1000.0, but then explicitly casted to int.
In this way x2 cannot assume values like 5.123
The resulting behavior is totally similar to the one obtained in the case above (light blue and red):
![image](https://user-images.githubusercontent.com/11663239/30035781-e8efde70-91ad-11e7-8ffb-5f7fb946a123.png)

I think that this could arise from the fact that I'm using python2 and therefore this could be solved by adding from __future__ import division in transformers.py
I'm now running a test with this modification.

**Comment by stefanocereda** [comment]
Yes, adding that fix solved the issue:

**Comment by stefanocereda** [comment]
The import fixed the issue:
![image](https://user-images.githubusercontent.com/11663239/30051636-0df4da7a-9223-11e7-88b5-334d80ca7a9b.png)
Can somebody make the proposed modification?

**Comment by glouppe** [comment]
Hi,

Thanks for your investigation. This is very interesting!

In summary, what do you propose to change? Note that we do have a normalization mechanism in skopt, but I think it is turned off by default. See https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/space/space.py#L169

Gilles

**Comment by stefanocereda** [comment]
If I remember the code correctly, the normalization is turned on when using gp_minimize().
The problem is that in transformers.py the value is normalized as (value - min) / (max-min) where by max and min I mean the variable's domain. When the domain is integer, the denominator is integer and python2 cast to integer the division result, leading to bad results.

This can be fixed by adding a `from __future__ import division` in the first line of transformers.py so that the division in python2 behaves like in python3 and returns a floating value.

**Comment by glouppe** [comment]
Oh okay, I see, this is because you are using Python 2. That's an easy fix and we should do that!

Are you willing to make a small contribution to scikit-optimize with that fix? :)

**Comment by stefanocereda** [comment]
Yes unfortunately I cannot decide which python to use on the machine I'm using.
Shall I clone the repo and make a pull request to make the fix?

**Comment by glouppe** [comment]
Yes.

1) Make a clone
2) Create a branch
3) Make the fix in that branch
4) Submit a PR from that branch to our repo.

**Comment by glouppe** [comment]
Fixed by #504
